### ISSUE TITLE
Overloads with different number of union members seem to give spurious overlap error

### ISSUE THREAD

**Issue by tibbe** [python/mypy#16545]
**Bug Report**

It seems to me that the code below should be accepted as a union of three members should be distinct from a union of two members. This is subtle so I might be missing something e.g when it comes to either how `TypeVar`s interact with unions (e.g. will a `TypeVar` match any union).

**To Reproduce**

```python
from typing import TypeVar, overload

T1 = TypeVar('T1')
T2 = TypeVar('T2')
T3 = TypeVar('T3')

@overload
def parse_as(ty: type[T1] | type[T2], data: bytes) -> T1 | T2:
    ...

@overload
def parse_as(ty: type[T1] | type[T2] | type[T3], data: bytes) -> T1 | T2 | T3:
    ...
```

**Expected Behavior**

I expected no type error.

**Actual Behavior**

Error: "Overloaded function signature 2 will never be matched: signature 1's parameter type(s) are the same or broader"

**Your Environment**

- Mypy version used: 1.5.1
- Mypy command-line flags: mypy repro.py
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.11

**Comment by tibbe** [comment]
I also tried this slight variation, with the same result:

```py
@overload
def parse_as2(ty: type[T1 | T2], data: bytes) -> T1 | T2:
    ...

@overload
def parse_as2(ty: type[T1 | T2 | T3], data: bytes) -> T1 | T2 | T3:
    ...
```

**Comment by JelleZijlstra** [comment]
I think mypy is right here. I don't know what you're trying to achieve by using multiple TypeVars and using them only in unions: how is this signature different from just `def parse_as(ty: type[T], data: bytes) -> T: ...`.

**Comment by tibbe** [comment]
I'm trying to work around `type`s inability to capture unions and do type inference on them (this is what the `TypeForm` proposal is about but that's not here yet).

**Comment by erictraut** [comment]
This should work today. No need for `TypeForm`.

```python
from typing import TypeVar

T = TypeVar("T")

def parse_as(ty: type[T], data: bytes) -> T:
    ...

def test(x: type[int | str]):
    parse_as(x, b"")
```

**Comment by tibbe** [comment]
Here's something closer to my real code:

```py
from typing import TypeAlias, TypeVar

T = TypeVar('T')

def parse_as(ty: type[T], data: bytes) -> T:
    ...

SomeUnion: TypeAlias = int | str

def test() -> SomeUnion:
    return parse_as(SomeUnion, b'')
```

Error: Argument 1 to "parse_as" has incompatible type "<typing special form>"; expected "type[<nothing>]"

**Comment by erictraut** [comment]
@tibbe, you're using a type alias inappropriately here. Type aliases should be used in type (annotation) expressions. They are not generally meant to be used in value (runtime) expressions. There are a few special cases where the runtime specifically allows type aliases to be used: a `bound` expression in a `TypeVar` definition, constrained types in a constrained `TypeVar` definition, or the second argument to an `isinstance` or `issubclass` call.

In your example, you're using the type alias as a value (runtime) expression, so mypy is treating it as a `UnionType` object (the runtime implementation of a union operator) rather than a union type.

You could fix this issue in your code by changing `parse_as` to accept a tuple of types, as follows:

```python
from typing import TypeAlias, TypeVar

T = TypeVar("T")

def parse_as(ty: tuple[type[T], ...], data: bytes) -> T:
    ...

SomeUnion: TypeAlias = int | str

def test() -> SomeUnion:
    return parse_as((int, str), b"")
```

This approach should work fine, but unfortunately it runs into a well-known problem with mypy where its constraint solver uses joins rather than unions. It therefore evaluates the call to `parse_as` as type `type[object]` rather than `type[int] | type[str]`. A false positive error is then reported for the return statement in `test`.

**Comment by tibbe** [comment]
Thanks for the explanation @erictraut. It's also quite a bit unergonomic unfortunately. We have many unions (defined using type aliases) in our code base. Having to have a tuple version of each union for the purpose of passing to `parse_as` is rather cumbersome:

```py
SomeUnion: TypeAlias = int | str
SomeTuple = (int, str)

def test() -> SomeUnion:
    return parse_as(SomeTuple , b"")
```

(Does this even work? Will `SomeTuple` implicitly be a type alias and run afoul of the problem you mentioned above?)

I'm not sure how to design this in a way that doesn't run into one problem or another. At the end of the day we have lots of (discriminate) unions in our code base (that we write/read to/from DynamoDB) and we need a way to say "parse the data we got as this particular (discriminate) union".

**Comment by erictraut** [comment]
> Does this even work?

It works fine in [pyright](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgIIA2SAhgM4A0hJAalSAFBsFQC8jCxLEAAoARAREBKDmwAmxYFASsaxAPq0h8AFywArgnLEA2vH5GCAXQYA6W1agyqMKjoBGcGMRoSoAWgB8hFpsUKFQttbSAMqQxACqKEhgKDpE-BTUNDyYKPgAPlA0MOwxEMQE%2BobZQqgwDEUgUhxyCp5FQj4BUKXxicnBYVAgxDC6ICiKymoaPRUGxAyuIpJsQA), but it doesn't work in mypy because of the join vs. union issue.

> Will `SomeTuple` implicitly be a type alias and run afoul of the problem you mentioned above?

This won't be interpreted as a type alias because tuples are not allowed in a type expression. Type checkers will therefore treat `SomeTuple` as a normal variable, not a type alias.
